### DISCUSSION TITLE
Convenience functions and exposed metadata

### DISCUSSION THREAD

**Discussion by maximedion2** [LDeakin/zarrs#101] [Q&A] [answered]
Hey there! This is a really cool library! I was considering using it in a project I'm working on, but I have a couple questions, I skimmed through the docs but I'm not sure about 2 functionalities that would be quite useful on my end. 

- are there "convenience" functions on top of the manual chunk writes? What I mean is e.g. once a store is set up, with chunk and array shapes, codecs, etc... you'd pass such a function an array (maybe nested vectors) and it handles compression, chunking, sharding, etc... and just writes to the appropriate files in the store?
- I see that the metadata can be returned as a json object, but is there any way to access the various things encoded in the metadata as actual objects (maybe enums, vectors for shapes, etc...) that can be checked in code, used in branching statements, etc...? I assume those objects exist inside the code, but they might not be publicly exposed?

**Comment by LDeakin** [comment, marked as answer]
I think you might be looking for the following methods:
- [`Array::store_array_subset`](https://docs.rs/zarrs/latest/zarrs/array/struct.Array.html#method.store_array_subset)
- [`Array::store_array_subset_elements`](https://docs.rs/zarrs/latest/zarrs/array/struct.Array.html#method.store_array_subset_elements)
- [`Array::store_array_subset_ndarray`](https://docs.rs/zarrs/latest/zarrs/array/struct.Array.html#method.store_array_subset_ndarray)

These can store an arbitrary subset of an array, you don't have to think about chunks. They (and other `Array::store_*` methods, excluding `store_encoded_chunk`) take care of everything for you, compression, sharding, etc.

You can grab individual components of the metadata: see https://docs.rs/zarrs/latest/zarrs/array/struct.Array.html#immutable-array-metadata--properties and below. EDIT: also [`Array::metadata()`](https://docs.rs/zarrs/latest/zarrs/array/struct.Array.html#method.metadata) returns an enum [`ArrayMetadata`](https://docs.rs/zarrs/latest/zarrs/array/enum.ArrayMetadata.html) that can represent V2/V3 metadata. The variants are completely `pub`.

**Reply by maximedion2** [reply]
Great, that seems like what I need, thanks!
